### PR TITLE
Support filtering for Elastic types.

### LINKED ISSUE #121: Collect Elasticsearch diagnostics of a given cluster using ECK diagnostics
It helps if we can add feature to collect specific deployment diagnostics using ECK diagnostics instead of collecting all deployments diagnostics data.

### PR DESCRIPTION
**Pull request by naemono**
# Proposed Change

closes #121 
The following change will add filtering for Elastic custom resource types.

## Supported Types

`agent apm beat elasticsearch enterprisesearch kibana maps`

## Limitations
* ~~Only one filter is supported at a time~~
```
# supported
eck-diagnostics -r a -f "elasticsearch=mycluster"

# also supported
eck-diagnostics -r a -f "elasticsearch=mycluster" -f "kibana=mykb"
```

### Filtered resources

Only a certain number of Kubernetes resources support filtering when filtering by an Elastic custom resource.  Along with the named Elastic custom resource type, the following resources will be returned that are associated:

* ConfigMap
* ControllerRevision
* Deployment
* DaemonSet
* Endpoint
* Pod
* PersistentVolumeClaim
* Replicaset
* Service
* StatefulSet

The following resources are returned unfiltered:

* Event
* NetworkPolicy
* PersistentVolume
* ServiceAccount
* Secret (metadata only)

### CODE REVIEW

**Review by thbkrkr — COMMENTED** [review]
Looks good. I think it makes sense to also apply the filters to find the Elastic resources for which we get logs and run agent/stack diagnostics.

File: cmd/main.go [hunk 1]
```diff
@@ -47,6 +52,15 @@ func main() {
 	}
 }
 
+func validateFilters(_ *cobra.Command, _ []string) error {
```
Comment (pebrc) [inline cmd/main.go:55]: It's doing more than validating. It is doing a transformation. Maybe `parseFilters`?

File: internal/filters/filters.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+// source example:
+// []string{"name=mycluster", "type=elasticsearch"}
+func New(source []string) (Filter, error) {
+	filter := Filter{
+		source: source,
+	}
+	return filter.validate()
+}
+
+func (f Filter) validate() (Filter, error) {
```
Comment (pebrc) [inline internal/filters/filters.go:63]: naming nit: this is doing more than validation: parsing, creating not sure what is a better word

File: internal/filters/filters.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+)
+
+// Filter is a type + name filter that translates into a labelSelector that is applied when querying for Kubernetes resources.
+// Both type and name are required at this time.
+//
+// examples supported:
+// name=mycluster, type=elasticsearch
+// name=mykb, type=kibana
+type Filter struct {
+	source        []string
```
Comment (pebrc) [inline internal/filters/filters.go:29]: Why do we need `source` in the struct when it is only used during creation?

File: internal/filters/filters.go (hunk 2 above)
Comment (pebrc) [reply internal/filters/filters.go:63]:
```suggestion
func parse(source []string) (Filter, error) {
```

I think it also does not need to be a receiver method of `Filter`

File: internal/filters/filters.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+// if the given key=value pairs are invalid.
+//
+// source example:
+// []string{"name=mycluster", "type=elasticsearch"}
+func New(source []string) (Filter, error) {
+	filter := Filter{
+		source: source,
+	}
+	return filter.validate()
+}
```
Comment (pebrc) [inline internal/filters/filters.go:61]:
```suggestion
func New(source []string) (Filter, error) {	
	return parse(source)
}
```
or inline parse into `New`

File: internal/filters/filters.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	var elasticfilter string
+	elasticfilter += elasticTypeKey + "=" + strings.ToLower(typ) + ","
+	switch typ {
+	case "elasticsearch":
+		elasticfilter += fmt.Sprintf(elasticsearchNameFormat, strings.ToLower(typ)) + "=" + name
+	default:
+		elasticfilter += fmt.Sprintf(elasticNameFormat, strings.ToLower(typ)) + "=" + name
+	}
+
+	return elasticfilter
```
Comment (pebrc) [inline internal/filters/filters.go:133]: Could this be simplified to use just one (or two maybe for clarity) Sprintfs, no concatenation, and no `ToLower` (because we already validated that earlier)?

```golang
format := "common.k8s.elastic.co/type=%s,%s.k8s.elastic.co/%s=%s"
	nameAttr := "name"
	if typ == "elasticsearch" {
		nameAttr = "cluster-name"
	}
	return fmt.Sprintf(format, typ, typ, nameAttr, name)
```

File: internal/kubectl.go [hunk 6]
```diff
@@ -177,6 +179,64 @@ func (c Kubectl) Get(resource, namespace string, w io.Writer) error {
...
+
+	obj, err := r.Object()
+	if err != nil {
+		return err
+	}
+
+	return printer.PrintObj(obj, w)
+}
+
+func extractTypeName(selectors string) (string, string, error) {
```
Comment (pebrc) [inline internal/kubectl.go:220]: only used in test

File: internal/filters/filters.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+// Type returns the type of the Elastic resources to filter.
+func (f Filter) Type() string {
+	return f.typ
+}
+
+// Name returns the name of the Elastic resource to filter.
+func (f Filter) Name() string {
+	return f.name
+}
```
Comment (pebrc) [inline internal/filters/filters.go:48]: I think these are redundant "Getter" style functions. Just export the struct fields? The struct is completely constructed at the end of the `New` constructor method and these function do no additional computation.

**Review by pebrc — COMMENTED** [review]
I left a few comments, all small things. One more fundamental concern I have is with the practicality of the chosen approach. Would we not typically want to have it so that we can extract diagnostics for an Elastic Stack deployment? E.g. Elasticsearch and Kibana, or Elasticsearch, Kibana and Agent etc. Of course sometimes we would want just the Elasticsearch diagnostics.  

I wonder how we could achieve that? 

1. we could lift the single filter restriction in the future. I am not sure how easy that would be with the chosen slice parameter type (how do you tell different filters apart) also the usability would be not great. Users having to specify complex filter expressions in a already tense situation when something is not working is maybe not the best experience.
2. another approach could be that we traverse `elasticsearchRefs` You specify a filter for Elasticsearch and filter all Elastic custom resource objects based on the presence of a elasticsearchRef pointing to the specified Elasticsearch resource

Both can probably be added in a follow up PR I just want to make sure we build the initial feature in such a way that is still possible to add later (also the chosen parameter name `filters` suggests that we can specify more than one). Maybe a more compact filter syntax would already help solve this last problem? `elasticsearch=my-cluster,kibana=my-kibana`

File: internal/kubectl.go (hunk 6 above)
Comment (naemono) [reply internal/kubectl.go]: Remnant of the previous implementation.  Thank you.  Removed

File: internal/filters/filters.go (hunk 7 above)
Comment (naemono) [reply internal/filters/filters.go]: I see your point. Exported, and removed.

**Comment by naemono** [comment]
> Maybe a more compact filter syntax would already help solve this last problem? `elasticsearch=my-cluster,kibana=my-kibana`

This certainly simplifies things.  Thank you.  I'll implement this immediately.  I'll comment on the other items once complete.

**Comment by naemono** [comment]
@pebrc @thbkrkr I've got a functional implementation now that supports multiple filters, but unfortunately doesn't parse the *Refs, as that seems to be a much more involved change that likely can come down the road if it's deemed necessary.

I'm cleaning up the implementation, and should have something submitted here shortly.

**Comment by naemono** [comment]
@pebrc @thbkrkr changes submitted, and this is ready for another set of 👀

**Comment by naemono** [comment]
> Looks good. I think it makes sense to also apply the filters to find the Elastic resources for which we get logs and run agent/stack diagnostics.

I didn't initially see this.  I'll handle this next.

File: internal/kubectl.go [hunk 8]
```diff
@@ -174,7 +177,77 @@ func (c Kubectl) Get(resource, namespace string, w io.Writer) error {
...
+		if err != nil {
+			return fmt.Errorf("while accessing name for %s: %w", item.Object.GetObjectKind().GroupVersionKind().String(), err)
+		}
+
+		if filters.Contains(name, resource) {
+			filtered.Items = append(filtered.Items, item)
+		}
+	}
+
+	return printer.PrintObj(&filtered, w)
```
Comment (pebrc) [inline internal/kubectl.go:250]: This function almost completely duplicates `Get` maybe we can remove some duplication?

```diff
diff --git a/internal/kubectl.go b/internal/kubectl.go
index 328147a..2453b75 100644
--- a/internal/kubectl.go
+++ b/internal/kubectl.go
@@ -163,50 +163,24 @@ func (c Kubectl) Exec(nsn types.NamespacedName, cmd ...string) error {
 // If filters is not empty, this will only return resources within the cluster that it's labels match
 // at least one of the filter's label selectors.
 func (c Kubectl) Get(resource, namespace string, filters internal_filters.Filters, w io.Writer) error {
-	r, err := c.getResources(resource, namespace)
-	if err != nil {
-		return err
-	}
-	printer, err := printers.NewTypeSetter(scheme.Scheme).WrapToPrinter(&printers.JSONPrinter{}, nil)
-	if err != nil {
-		return err
-	}
-
-	obj, err := r.Object()
-	if err != nil {
-		return err
-	}
-
-	// If there are no filters, simply return the unfiltered resources.
-	if filters.Empty() {
-		return printer.PrintObj(obj, w)
-	}
-
-	// Otherwise, convert the returned resource to a List, and filter for any matching objects
-	// that have labels that match any of the given filter's label selector.
-	var list *corev1.List
-	list, ok := obj.(*corev1.List)
-	if !ok {
-		return fmt.Errorf("while converting returned object (%T) to list", obj)
-	}
-
-	filtered := corev1.List{}
-	for _, item := range list.Items {
-		labels, err := meta.NewAccessor().Labels(item.Object)
-		if err != nil {
-			return fmt.Errorf("while accessing labels for %s: %w", item.Object.GetObjectKind().GroupVersionKind().String(), err)
-		}
-		if filters.Matches(labels) {
-			filtered.Items = append(filtered.Items, item)
-		}
-	}
-
-	return printer.PrintObj(&filtered, w)
+	return c.getFiltered(resource, namespace, w,
+		func(object metav1.Object) bool {
+			return filters.Matches(object.GetLabels())
+		},
+		filters.Empty())
 }
 
 // GetElastic retrieves the Elastic K8s objects of type resource in namespace,
 // filters by name, and marshals them into the writer w.
 func (c Kubectl) GetElastic(resource, namespace string, filters internal_filters.Filters, w io.Writer) error {
+	return c.getFiltered(resource, namespace, w,
+		func(object metav1.Object) bool {
+			return filters.Contains(object.GetName(), resource)
+		},
+		filters.Empty())
+}
+
+func (c Kubectl) getFiltered(resource, namespace string, w io.Writer, filter func(object metav1.Object) bool, skipFilter bool) error {
 	r, err := c.getResources(resource, namespace)
 	if err != nil {
 		return err
@@ -223,7 +197,7 @@ func (c Kubectl) GetElastic(resource, namespace string, filters internal_filters
 	}
 
 	// If there are no filters, simply return the unfiltered resources.
-	if filters.Empty() {
+	if skipFilter {
 		return printer.PrintObj(obj, w)
 	}
 
@@ -237,12 +211,12 @@ func (c Kubectl) GetElastic(resource, namespace string, filters internal_filters
 
 	filtered := corev1.List{}
 	for _, item := range list.Items {
-		name, err := meta.NewAccessor().Name(item.Object)
+		obj, err := meta.Accessor(item.Object)
 		if err != nil {
-			return fmt.Errorf("while accessing name for %s: %w", item.Object.GetObjectKind().GroupVersionKind().String(), err)
+			return fmt.Errorf("while accessing metadata for %s: %w", item.Object.GetObjectKind().GroupVersionKind().String(), err)
 		}
 
-		if filters.Contains(name, resource) {
+		if filter(obj) {
 			filtered.Items = append(filtered.Items, item)
 		}
 	}
```

**Review by pebrc — COMMENTED** [review]
Almost LGTM, a few small things. Tested it and it worked as expected.

File: internal/filters/filters.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+func processSelector(typ, name string) (labels.Selector, error) {
+	nameAttr := "name"
+	if typ == "elasticsearch" {
+		nameAttr = "cluster-name"
+	}
+	set := labels.Set{
+		"common.k8s.elastic.co/type":                       typ,
+		fmt.Sprintf("%s.k8s.elastic.co/%s", typ, nameAttr): name,
+	}
+	s := labels.SelectorFromSet(set)
```
[the remark this answers is not in the document]
Comment (naemono) [reply internal/filters/filters.go]: Yes we could.  Updated.

File: internal/kubectl.go [hunk 10]
```diff
@@ -328,11 +401,19 @@ func (c Kubectl) Logs(namespace string, selector string, out func(string) (io.Wr
...
+				if !filters.Empty() && !filters.Matches(p.Labels) {
+					continue
+				}
 				if err := c.requestLogs(p, out); err != nil {
 					return err
 				}
 			}
 		case *corev1.Pod:
+			// ignore pod when filters are being applied, and the pod doesn't match the filters.
+			if !filters.Empty() && !filters.Matches(t.Labels) {
```
[the remark this answers is not in the document]
Comment (naemono) [reply internal/kubectl.go]: Makes perfect sense.  Done

File: internal/filters/filters.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+var (
+	// ValidTypes are the valid types of Elastic resources that are supported by the filtering system.
+	ValidTypes = []string{"agent", "apm", "beat", "elasticsearch", "enterprisesearch", "kibana", "maps"}
+)
+
+// Filters contains a Filter map for each Elastic type given in the filter "source".
+type Filters struct {
+	ByType map[string][]Filter
```
[the remark this answers is not in the document]
Comment (naemono) [reply internal/filters/filters.go]: Good point.  Un-exported.

File: internal/kubectl.go [hunk 12]
```diff
@@ -174,7 +177,77 @@ func (c Kubectl) Get(resource, namespace string, w io.Writer) error {
...
+			filtered.Items = append(filtered.Items, item)
+		}
+	}
+
+	return printer.PrintObj(&filtered, w)
+}
+
+// GetElastic retrieves the Elastic K8s objects of type resource in namespace,
+// filters by name, and marshals them into the writer w.
+func (c Kubectl) GetElastic(resource, namespace string, filters internal_filters.Filters, w io.Writer) error {
```
[the remark this answers is not in the document]
Comment (naemono) [reply internal/kubectl.go]: Very true.  Updated.

File: internal/kubectl.go (hunk 8 above)
Comment (naemono) [reply internal/kubectl.go:225]: That certainly DRYs things up a bit.  Thank you.

File: internal/filters/filters.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+	if typ == "elasticsearch" {
+		nameAttr = "cluster-name"
+	}
+	set := labels.Set{
+		"common.k8s.elastic.co/type":                       typ,
+		fmt.Sprintf("%s.k8s.elastic.co/%s", typ, nameAttr): name,
+	}
+	s := labels.SelectorFromValidatedSet(set)
+	for sk, v := range set {
+		req, err := labels.NewRequirement(sk, selection.Equals, []string{v})
```
Comment (pebrc) [inline internal/filters/filters.go:149]: `SelectorFromValidatedSet` already adds the Requirement, that's why I recommended using it.

Comment (naemono) [reply internal/filters/filters.go]: Going and looking at the internal implementation, I see this now.  Thank you.  Interesting that `labels.SelectorFromSet(set)` uses `SelectorFromValidatedSet` internally, so is functionally equivalent.

File: internal/filters/filters_test.go [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+			}},
+		},
+	}
+	tests := []struct {
+		name      string
+		labels    map[string]string
+		filterMap map[string][]Filter
+		want      bool
+	}{
+		{
```
Comment (pebrc) [inline internal/filters/filters_test.go:211]: add test for empty filters?

File: internal/stackdiag.go [hunk 15]
```diff
@@ -394,6 +395,10 @@ func scheduleJobs(k *Kubectl, ns string, recordErr func(error), state *diagJobSt
 		}
 		tls := !(found && disabled)
 
+		if !filters.Contains(resourceName, typ) {
```
Comment (pebrc) [inline internal/stackdiag.go:398]: I think you still need an empty check otherwise you disable stack diagnostics when no filters are defined.

**Review by pebrc — COMMENTED** [review]
I think we have a small regression now in the stack diagnostics code. See comments. Otherwise looks good.

File: internal/stackdiag.go (hunk 15 above)
Comment (naemono) [reply internal/stackdiag.go]: Thanks for seeing this.  Added this back.
